### ISSUE TITLE
Flutter failed to run iOS project with Swift Macro on iOS Simulator

### ISSUE THREAD

**Issue by codingiran** [flutter/flutter#146122]
### Type of Request

bug

### Infrastructure Environment

My develop env is 
```
Doctor summary (to see all details, run flutter doctor -v):
[✓] Flutter (Channel stable, 3.19.5, on macOS 14.4.1 23E224 darwin-arm64, locale zh-Hans-CN)
[✓] Android toolchain - develop for Android devices (Android SDK version 34.0.0)
[✓] Xcode - develop for iOS and macOS (Xcode 15.3)
[✓] Chrome - develop for the web
[✓] Android Studio (version 2023.2)
[✓] VS Code (version 1.87.2)
[✓] Connected device (4 available)
    ! Error: Browsing on the local area network for iPhone. Ensure the device is unlocked and attached with a cable or associated with the same local area network as this Mac.
      The device must be opted into Developer Mode to connect wirelessly. (code -27)
    ! Error: Browsing on the local area network for iPhone. Ensure the device is unlocked and attached with a cable or associated with the same local area network as this Mac.
      The device must be opted into Developer Mode to connect wirelessly. (code -27)
    ! Error: Browsing on the local area network for 客厅’s Apple TV. Ensure the device is unlocked and associated with the same local area network as this Mac. (code -27)
    ! Error: Browsing on the local area network for Hai‘s phone. Ensure the device is unlocked and attached with a cable or associated with the same local area network as this Mac.
      The device must be opted into Developer Mode to connect wirelessly. (code -27)
[✓] Network resources

• No issues found!

```

### What is happening?

[Marco](https://docs.swift.org/swift-book/documentation/the-swift-programming-language/macros/) comes in lastest Swift 

When a Flutter project using Swift Package Manager to import a Libary with Marco, Flutter build ios failed!

```
Launching lib/main.dart on iPhone 15 Pro in debug mode...
Running Xcode build...
Xcode build done.                                           33.6s
Failed to build iOS app
Swift Compiler Error (Xcode): External macro implementation type 'MacroPlugin.Codable' could not be found for macro 'Codable()'
/Users/codingiran/Downloads/my_first_flutter_project/ios/Runner/AppDelegate.swift:17:7

Swift Compiler Error (Xcode): External macro implementation type 'MacroPlugin.CodedAt' could not be found for macro 'CodedAt'
/Users/codingiran/Downloads/my_first_flutter_project/ios/Runner/AppDelegate.swift:19:8

Could not build the application for the simulator.
```

### Steps to reproduce

Using minimal reproduce flutter proj:

[my_first_flutter_project.zip](https://github.com/flutter/flutter/files/14834817/my_first_flutter_project.zip)

### Expected results

I expect to run in iOS simulator of Xcode 15.3

**Comment by codingiran** [comment]
Related:

https://forums.developer.apple.com/forums/thread/737908
https://stackoverflow.com/questions/77302264/use-skipmacrovalidation-for-xcodebuild-command-when-using-an-external-package
https://forums.swift.org/t/is-there-a-way-to-programmatically-allow-trust-the-compiler-plugin-to-run-from-the-command-line/65690

**Comment by huycozy** [comment]
Thanks for the report @codingiran. I can reproduce the issue on my lower XCode version 15.2 as well (running on iPhone 15 Pro, iOS 17.2 emulator).

<details>
<summary>flutter doctor -v (stable and master)</summary>

```bash
[✓] Flutter (Channel stable, 3.19.5, on macOS 14.1 23B74 darwin-x64, locale en-VN)
    • Flutter version 3.19.5 on channel stable at /Users/huynq/Documents/GitHub/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 300451adae (25 hours ago), 2024-03-27 21:54:07 -0500
    • Engine revision e76c956498
    • Dart version 3.3.3
    • DevTools version 2.31.1

[✓] Android toolchain - develop for Android devices (Android SDK version 34.0.0)
    • Android SDK at /Users/huynq/Library/Android/sdk
    • Platform android-34, build-tools 34.0.0
    • ANDROID_HOME = /Users/huynq/Library/Android/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jbr/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 17.0.9+0-17.0.9b1087.7-11185874)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 15.2)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 15C500b
    • CocoaPods version 1.15.2

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2023.2)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • android-studio-dir = /Applications/Android Studio.app/
    • Java version OpenJDK Runtime Environment (build 17.0.9+0-17.0.9b1087.7-11185874)

[✓] VS Code (version 1.87.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.84.0

[✓] Connected device (3 available)
    • Pixel 7 (mobile) • 2B171FDH20084L • android-arm64  • Android 14 (API 34)
    • macOS (desktop)  • macos          • darwin-x64     • macOS 14.1 23B74 darwin-x64
    • Chrome (web)     • chrome         • web-javascript • Google Chrome 123.0.6312.86

[✓] Network resources
    • All expected network resources are available.

• No issues found!
```

```bash
[!] Flutter (Channel master, 3.22.0-2.0.pre.19, on macOS 14.1 23B74 darwin-x64, locale en-VN)
    • Flutter version 3.22.0-2.0.pre.19 on channel master at /Users/huynq/Documents/GitHub/flutter_master
    ! Warning: `flutter` on your path resolves to /Users/huynq/Documents/GitHub/flutter/bin/flutter, which is not inside your current Flutter SDK checkout at /Users/huynq/Documents/GitHub/flutter_master. Consider adding /Users/huynq/Documents/GitHub/flutter_master/bin to the front of your path.
    ! Warning: `dart` on your path resolves to /Users/huynq/Documents/GitHub/flutter/bin/dart, which is not inside your current Flutter SDK checkout at /Users/huynq/Documents/GitHub/flutter_master. Consider adding /Users/huynq/Documents/GitHub/flutter_master/bin to the front of your path.
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 0ee25923ae (2 hours ago), 2024-04-01 21:22:29 -0400
    • Engine revision 9a513de652
    • Dart version 3.5.0 (build 3.5.0-6.0.dev)
    • DevTools version 2.34.1
    • If those were intentional, you can disregard the above warnings; however it is recommended to use "git" directly to perform update checks and upgrades.

[✓] Android toolchain - develop for Android devices (Android SDK version 34.0.0)
    • Android SDK at /Users/huynq/Library/Android/sdk
    • Platform android-34, build-tools 34.0.0
    • ANDROID_HOME = /Users/huynq/Library/Android/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jbr/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 17.0.9+0-17.0.9b1087.7-11185874)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 15.2)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 15C500b
    • CocoaPods version 1.15.2

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2023.2)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • android-studio-dir = /Applications/Android Studio.app/
    • Java version OpenJDK Runtime Environment (build 17.0.9+0-17.0.9b1087.7-11185874)

[✓] VS Code (version 1.87.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.84.0

[✓] Connected device (3 available)
    • RMX2001 (mobile) • EUYTFEUSQSRGDA6D • android-arm64  • Android 11 (API 30)
    • macOS (desktop)  • macos            • darwin-x64     • macOS 14.1 23B74 darwin-x64
    • Chrome (web)     • chrome           • web-javascript • Google Chrome 123.0.6312.87
    ! Error: Browsing on the local area network for iPad. Ensure the device is unlocked and attached with a cable or associated with the same local area network as this Mac.
      The device must be opted into Developer Mode to connect wirelessly. (code -27)

[✓] Network resources
    • All expected network resources are available.

! Doctor found issues in 1 category.
```
</details>

**Comment by vashworth** [comment]
@codingiran Can you try running the project through Xcode (open ios/Runner.xcworkspace) and then run `flutter attach` to be able to use hot reload as a workaround?

**Comment by codingiran** [comment]
> @codingiran Can you try running the project through Xcode (open ios/Runner.xcworkspace) and then run `flutter attach` to be able to use hot reload as a workaround?

YES, Run directly in Xcode and using `Flutter attach` is all going well.

**Comment by jmagman** [comment]
When I build for the simulator, in verbose mode I see "warning: compiler plugin not loaded: 'MacroPlugin; failed to initialize"
```
SwiftEmitModule normal arm64 Emitting\ module\ for\ Runner (in target 'Runner' from project 'Runner')
...
<unknown>:0: warning: compiler plugin not loaded: 'my_first_flutter_project/build/ios/Debug-iphonesimulator/MacroPlugin; failed to initialize
my_first_flutter_project/ios/Runner/AppDelegate.swift:18:8: error: external macro implementation type 'MacroPlugin.Codable' could not be found for macro 'Codable()'
                    struct Test {
```

Very weirdly when targeting iOS, on both the command line xcodebuild targeting the physical phone and from Xcode when targeting the simulator, `MacroPlugin` is correctly being emitted for macOS `-target arm64-apple-macos10.15`.  However, on command line xcodebuild when targeting the simulator it's incorrectly building the plugin for the simulator `-target arm64-apple-ios13.0-simulator`.


Verbose output: [module_plugin.txt](https://github.com/flutter/flutter/files/14861732/module_plugin.txt)

I tried to make a new vanilla Xcode project with a dependency `MetaCodable` and 
```swift
@Codable
struct Test {
    @CodedAt("my_path")
    var myPath: String?
}
```
but it crashes xcodebuild with a SIGTRAP.

This looks like the same issue, also specific to `MetaCodable`
https://stackoverflow.com/questions/77386744/swift-macros-external-macro-implementation-type-could-not-be-found

This is either an issue with `MetaCodable` specifically (maybe it's missing something in its plugin definition), or it's a Swift Package bug.  It doesn't make sense that running `xcodebuild -sdk iphonesimulator` would fail but running from Xcode targeting a simulator would succeed.

I suggest starting by filing an issue with `MetaCodable` with a non-Flutter reproduction case, and also filing an issue with https://github.com/apple/swift-package-manager/issues/

This might even already be fixed in Swift 5.10, if https://github.com/apple/swift-package-manager/issues/6673 is reporting the same thing: https://github.com/apple/swift-package-manager/commit/1daaa1c0339584e50d42f62b5a7794a2e97b5782 

Closing, there wouldn't be anything Flutter can do about this.

**Comment by jmagman** [comment]
Listed in Xcode 16 resolved issues:
Swift packages containing macros will fail to build when cross-compiling with swift build --swift-sdk (105991372)

https://developer.apple.com/documentation/xcode-release-notes/xcode-16-release-notes#Swift-Packages
